### PR TITLE
Allows writing text with crayons

### PR DESCRIPTION
**Pull request by unid15**
15 characters limit. Font is comic sans MS

### CODE REVIEW

**Comment by Probe1** [comment]
DONT HONK
CLOWN INSIDE

It could use a font that looks more hand written but honestly i'm not hung up on it, it looks fine.

**Comment by ghost** [comment]
> maximum 10 letters
> 15 character limit

WHO DO I BELIEVE

**Comment by unid15** [comment]
@DrCelt the gif is a bit outdated, it's 15

**Comment by 9600bauds** [comment]
not entirely sure about this one
Any way to make it into a more handwriting-like font?

**Comment by ghost** [comment]
@9600bauds not using maptext, I don't think.

**Comment by 9600bauds** [comment]
Seems like it allows for HTML, try giving it the comic sans span

**Comment by Ephx1** [comment]
> DON'T DEAD
> OPEN INSIDE

that doesn't even make any sense you dummy!

**Comment by clusterfack** [comment]
Huh never knew you could use styles with maptext.

God byond is shittily documented.

**Comment by Intigracy** [comment]
What happens if I'm a faggot and do:

D
D
D
D
D
D
D

14 chars

File: code/game/objects/items/crayons.dm [hunk 1]
```diff
@@ -123,17 +124,37 @@ var/global/list/all_graffitis = list(
...
 			if("rune")
-				to_chat(user, "You start drawing a rune on the [target.name].")
+				to_chat(user, "You start drawing a rune on \the [target].")
+			if("text")
+				#define MAX_LETTERS 15
+				preference = input("Write some text here (maximum [MAX_LETTERS] letters).", "Crayon scribbles") as null|text
+				preference = copytext(preference, 1, MAX_LETTERS)
+				#undef MAX_LETTERS
+
+				if(length(replacetext(preference, " ", "")) == 0) //If there is no text
```
Comment (ComicIronic) [inline code/game/objects/items/crayons.dm]: Does replacetext work for the null?

Comment (unid15) [inline code/game/objects/items/crayons.dm]: Yep

**Comment by unid15** [comment]
I don't really like how comic sans looks here. I'll see what other fonts look good

**Comment by 9600bauds** [comment]
At least it works!

**Comment by ghost** [comment]
Maaaan maptext is cooler than I thought.

**Comment by unid15** [comment]
Actually it doesn't look that bad compared to other fonts I tried



A bit hard to read, but we can always change the font to something else

**Comment by 9600bauds** [comment]
You could even bump the font size up by one. Actually can you also center it?

**Comment by Shadowmech88** [comment]
I, too, would like it centered.

**Comment by unid15** [comment]
Added options to align the text to the top, middle or bottom. Also a fancy preview

**Comment by ComicIronic** [comment]
Don't hard del that image, the GC will get it.

**Comment by Probe1** [comment]
looks nice

File: code/game/objects/items/crayons.dm [hunk 2]
```diff
@@ -123,17 +128,59 @@ var/global/list/all_graffitis = list(
...
+					I.maptext = {"<span style="color:[colour];font-size:[FONT_SIZE];font-family:'[FONT_NAME]';" valign="[alignment]">[preference]</span>"}
+					I.loc = get_turf(target)
+					animate(I, alpha = 100, 10, -1)
+					animate(alpha = 255, 10, -1)
+
+					user.client.images.Add(I)
+					var/continue_drawing = alert(user, "This is how your drawing will look. Continue?", "Crayon scribbles", "Yes", "Cancel")
+
+					user.client.images.Remove(I)
+					qdel(I)
```
Comment (clusterfack) [inline code/game/objects/items/crayons.dm]: You can't actually qdel() an image. You just remove all the references to it, as well as cancel the animation in this case by doing animate(whatever).

File: code/game/objects/items/crayons.dm [hunk 3]
```diff
@@ -98,12 +98,17 @@ var/global/list/all_graffitis = list(
...
 	return
 
+#define FONT_SIZE "6pt"
+#define FONT_NAME "Comic Sans MS"
 /obj/item/toy/crayon/afterattack(atom/target, mob/user as mob, proximity)
 	if(!proximity) return
 
 	if(istype(target, /turf/simulated))
-		var/drawtype = input("Choose what you'd like to draw.", "Crayon scribbles") in list("graffiti","rune","letter")
+		var/drawtype = input("Choose what you'd like to draw.", "Crayon scribbles") in list("graffiti","rune","letter","text")
```
Comment (PJB3005) [inline code/game/objects/items/crayons.dm:107]: Oh god this line.

**Comment by clusterfack** [comment]
Other than that I cant find any issues

File: code/game/objects/items/crayons.dm (hunk 2 above)
Comment (unid15) [inline code/game/objects/items/crayons.dm]: Done

File: code/game/objects/items/crayons.dm [hunk 4]
```diff
@@ -156,7 +156,8 @@ var/global/list/all_graffitis = list(
 					var/continue_drawing = alert(user, "This is how your drawing will look. Continue?", "Crayon scribbles", "Yes", "Cancel")
 
 					user.client.images.Remove(I)
-					qdel(I)
+					animate(I) //Cancel the animation so that the image gets garbage collected
+					I = null
```
Comment (clusterfack) [inline code/game/objects/items/crayons.dm]: Almost but ```I.loc = null``` instead

**Comment by kol1th** [comment]
oh boy here we go

**Pull request merged**